src/cleanup_guard.rs

```diff
@@ -0,0 +1,120 @@
+use std::sync::Mutex;
+use std::{io, thread};
+
+use lazy_static::lazy_static;
+use slab::Slab;
+
+lazy_static! {
+    /// Contains the callbacks passed to currently-live [`CleanupGuard`]s
+    static ref LIVE_GUARDS: Mutex<GuardTable> = Mutex::new(Slab::new());
+}
+
+type GuardTable = Slab<Box<dyn FnOnce() + Send>>;
+
+/// Prepare to run [`CleanupGuard`]s on `SIGINT`/`SIGTERM`
+pub fn init() {
+    if let Err(e) = platform::init() {
+        eprintln!("couldn't register signal handler: {}", e);
+    }
+}
+
+/// A drop guard that also runs on `SIGINT`/`SIGTERM`
+// Could be extended to not require `'static` at the cost of significantly more unsafe.
+pub struct CleanupGuard {
+    slot: usize,
+}
+
+impl CleanupGuard {
+    /// Invoke `f` when dropped or killed by `SIGINT`/`SIGTERM`
+    pub fn new<F: FnOnce() + Send + 'static>(f: F) -> Self {
+        let guards = &mut *LIVE_GUARDS.lock().unwrap();
+        Self {
+            slot: guards.insert(Box::new(f)),
+        }
+    }
+}
+
+impl Drop for CleanupGuard {
+    fn drop(&mut self) {
+        let guards = &mut *LIVE_GUARDS.lock().unwrap();
+        let f = guards.remove(self.slot);
+        f();
+    }
+}
+
+// Invoked on a background thread. Process exits after this returns.
+fn on_signal(guards: &mut GuardTable) {
+    for guard in guards.drain() {
+        guard();
+    }
+}
+
+#[cfg(unix)]
+mod platform {
+    use std::os::unix::io::{IntoRawFd as _, RawFd};
+    use std::os::unix::net::UnixDatagram;
+    use std::panic::AssertUnwindSafe;
+
+    use libc::{c_int, SIGINT, SIGTERM};
+
+    use super::*;
+
+    pub fn init() -> io::Result<()> {
+        let (send, recv) = UnixDatagram::pair()?;
+
+        // Spawn a background thread that waits for the signal handler to write a signal
+        // into it
+        thread::spawn(move || {
+            let mut buf = [0];
+            let signal = match recv.recv(&mut buf) {
+                Ok(1) => c_int::from(buf[0]),
+                _ => unreachable!(),
+            };
+            // We must hold the lock for the remainder of the process's lifetime to avoid a
+            // race where a guard is created between `on_signal` and `raise`.
+            let guards = &mut *LIVE_GUARDS.lock().unwrap();
+            if let Err(e) = std::panic::catch_unwind(AssertUnwindSafe(|| on_signal(guards))) {
+                match e.downcast::<String>() {
+                    Ok(s) => eprintln!("signal handler panicked: {}", s),
+                    Err(_) => eprintln!("signal handler panicked"),
+                }
+            }
+            unsafe {
+                libc::signal(signal, libc::SIG_DFL);
+                libc::raise(signal);
```

Comment:
I think `std::process::exit(some_code_dedicated_for_interrupt)` can be used instead of killing the process asynchronously.

That probably allows us to use some signal abstraction crate. I don't know if any of us can write Windows version of this function.

Reply:
> I think std::process::exit(some_code_dedicated_for_interrupt) can be used

Unfortunately the difference between calling `exit` and terminating due to a signal is observable via e.g. https://www.gnu.org/software/libc/manual/html_node/Process-Completion-Status.html and confusing them may make it difficult to to compose gracefully in e.g. shell scripts. This is also why I opted not to use the `ctrlc` crate.

>instead of killing the process asynchronously.

Per https://pubs.opengroup.org/onlinepubs/009695399/functions/raise.html, `raise` is not asynchronous:
> If a signal handler is called, the raise() function shall not return until after the signal handler does.

> I don't know if any of us can write Windows version of this function.

https://learn.microsoft.com/en-us/windows/console/registering-a-control-handler-function looks pretty straightforward to me; I'm prepared to draft something based on that in future work.

Reply:
> the difference between calling exit and terminating due to a signal is observable

Yeah, but do we need to copy the exact behavior of signal termination? 
Personally, I'm fine that `hg` exits with 255 or `cargo nextest` with 100 on ^C.

> raise is not asynchronous:

Right, thanks.

Reply:
> Yeah, but do we need to copy the exact behavior of signal termination?

I think we shouldn't be gratuitously incompatible with standard conventions when they exist.

Reply:
> I think we shouldn't be gratuitously incompatible with standard conventions when they exist.

True. If we could leverage `ctrlc` crate for example, that can be a valid reason to slightly deviate from the standard behavior. That's the point.

Reply:
I think the modest amount of backend code here is a small price to pay to improve composability.

Reply:
I'm not keen on maintaining low-level signal handling code, but that's not a blocker.

Reply:
I would be very surprised if the low level code here needs to get touched more than a few times in the next decade. Worst-case scenario, it's easy to replace, since all we need to do is run some code soon-ish after a signal is received and then (probably) terminate the process somehow.